pkg/restic/repository_manager.go

```diff
@@ -0,0 +1,404 @@
+	rm := &repositoryManager{
+		objectStore:   objectStore,
+		bucket:        config.ResticLocation,
+		secretsClient: secretsClient,
+		log:           log,
+	}
+
+	switch BackendType(config.Name) {
+	case AWSBackend:
+		url := "s3.amazonaws.com"
+
+		// non-AWS, S3-compatible object store
+		if config.Config != nil && config.Config["s3Url"] != "" {
+			url = config.Config["s3Url"]
+		}
+
+		rm.repoPrefix = fmt.Sprintf("s3:%s/%s", url, rm.bucket)
+	case AzureBackend:
+		rm.repoPrefix = "azure:" + rm.bucket + ":"
+	case GCPBackend:
+		rm.repoPrefix = "gs:" + rm.bucket + ":"
+	}
+
+	return rm
+}
+
+func (rm *repositoryManager) RepoPrefix() string {
+	return rm.repoPrefix
+}
+
+func (rm *repositoryManager) RepositoryExists(name string) (bool, error) {
+	repos, err := rm.getAllRepos()
+	if err != nil {
+		return false, err
+	}
+
+	for _, repo := range repos {
+		if repo == name {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (rm *repositoryManager) InitRepo(name string) error {
+	resticCreds, err := rm.secretsClient.Get(credsSecret, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		secret := &corev1api.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: credsSecret,
+			},
+			Type: corev1api.SecretTypeOpaque,
+		}
+		if resticCreds, err = rm.secretsClient.Create(secret); err != nil {
+			return errors.WithStack(err)
+		}
+	} else if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// do we already have a key for this repo? we shouldn't
+	if _, exists := resticCreds.Data[name]; exists {
+		return errors.New("restic-credentials secret already contains an encryption key for this repo")
+	}
+
+	// generate an encryption key for the repo
+	key := make([]byte, 256)
+	if _, err := rand.Read(key); err != nil {
+		return errors.Wrap(err, "unable to generate encryption key")
+	}
+
+	// capture current state for patch generation
+	preBytes, err := json.Marshal(resticCreds)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if resticCreds.Data == nil {
+		resticCreds.Data = make(map[string][]byte)
+	}
+
+	// TODO dev code only
+	resticCreds.Data[name] = []byte("passw0rd")
+
+	// get the modified state and generate a patch
+	postBytes, err := json.Marshal(resticCreds)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	patch, err := jsonpatch.CreateMergePatch(preBytes, postBytes)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// patch the secret
+	if _, err := rm.secretsClient.Patch(resticCreds.Name, types.MergePatchType, patch); err != nil {
+		return errors.Wrap(err, "unable to patch restic-credentials secret")
+	}
+
+	// init the repo
+	cmd := &command{
+		baseName:   "/restic",
+		command:    "init",
+		repoPrefix: rm.repoPrefix,
+		repo:       name,
+	}
+
+	return errorOnly(rm.exec(cmd))
+}
+
+func (rm *repositoryManager) getAllRepos() ([]string, error) {
+	prefixes, err := rm.objectStore.ListCommonPrefixes(rm.bucket, "/")
```

Comment:
If the resticLocation that came from the config was bucket/dir, that goes into rm.bucket - will that work here having both the bucket and the directory in a single string? Or do we need to call ListCommonPrefixes(bucketName, dir+"/")?